src/python/pants/backend/project_info/peek.py

```diff
@@ -0,0 +1,146 @@
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import json
+import os
+from dataclasses import asdict, is_dataclass
+from enum import Enum
+from itertools import repeat
+from typing import Iterable, Mapping, Tuple, TypeVar, Union, cast
+
+from pkg_resources import Requirement
+
+from pants.engine.addresses import Address, Addresses, BuildFileAddress
+from pants.engine.console import Console
+from pants.engine.fs import DigestContents, FileContent, PathGlobs
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.rules import Get, MultiGet, collect_rules, goal_rule
+from pants.engine.target import Target, Targets
+
+
+class OutputOptions(Enum):
+    RAW = "raw"
+    JSON = "json"
+
+
+class PeekSubsystem(GoalSubsystem):
+    """Display BUILD file info to the console.
+
+    In its most basic form, `peek` just prints the contents of a BUILD file. It can also display
+    multiple BUILD files, or render normalized target metadata as JSON for consumption by other
+    programs.
+    """
+
+    name = "peek"
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--output",
+            type=OutputOptions,
+            default=OutputOptions.RAW,
+            help="which output style peek should use",
+        )
+
+    @property
+    def output(self) -> OutputOptions:
+        return cast(OutputOptions, self.options.output)
+
+
+class Peek(Goal):
+    subsystem_cls = PeekSubsystem
+
+
+_T = TypeVar("_T")
+_U = TypeVar("_U")
+
+
+def _intersperse(x: _T, ys: Iterable[_U]) -> Iterable[Union[_T, _U]]:
+    # Use stateful iterator to safely iterate through first item
+    it = iter(ys)
+    for y in it:
+        yield y
+        break
+    for x, y in zip(repeat(x), it):
+        yield x
+        yield y
+
+
+def _render_raw(fc: FileContent, encoding: str = "utf-8") -> str:
+    dashes = "-" * len(fc.path)
+    content = fc.content.decode(encoding)
+    parts = [dashes, fc.path, dashes, content]
+    if not content.endswith(os.linesep):
+        parts.append("")
+    return os.linesep.join(parts)
+
+
+def _render_json(ts: Iterable[Target]) -> str:
+    data = dict(map(_target_to_kv, ts))
+    return json.dumps(data, indent=2, cls=_PeekJsonEncoder)
+
+
+def _target_to_kv(t: Target) -> Tuple[str, dict]:
+    nothing = object()
+    d = {
+        k.alias: v.value
+        for k, v in t.field_values.items()
+        # don't report default values in normalized output
+        if getattr(k, "default", nothing) != v.value
+    }
+    d["alias"] = t.alias
+    return t.address.spec, d
+
+
+class _PeekJsonEncoder(json.JSONEncoder):
+    """Allow us to serialize some commmonly-found types in BUILD files."""
+
+    safe_to_str_types = (Requirement,)
+
+    def default(self, o):
+        """Return a serializable object for o."""
+        if is_dataclass(o):
+            return asdict(o)
+        elif isinstance(o, Mapping):
```

Comment:
Another common one is `Sequence`.

Also, I think this `isinstance` check is supposed to use `collections.abc.Mapping`, rather than `typing`. To avoid ambiguity, I recommend using the fully specified `collections.abc.Mapping` instead of `from collections.abc import Mapping`.

Reply:
- [ ] add `Sequence` and switch to `collections.abc`